_posts/2023-04-19-#27039.md

```diff
@@ -7,20 +7,47 @@ authors: [pinheadmz]
+  - the `blocks/index/` block index data which is stored in a LevelDB database. The index helps to quickly find blocks without having to re-scan all the data files.
+    - At runtime, the block index is loaded into [`BlockManager::m_block_index`](https://github.com/bitcoin/bitcoin/blob/19764dc143281376ea08e954018479ed10405b72/src/node/blockstorage.h#L152). However, we also maintain statistics about the block files in [`BlockManager::m_blockfile_info`](https://github.com/bitcoin/bitcoin/blob/19764dc143281376ea08e954018479ed10405b72/src/node/blockstorage.h#L117).
+
+- Flushing is the process of bringing the data that we keep in-memory in sync with the data we store on disk. We want to flush regularly enough to ensure that we can fully recover when bitcoind is unexpectedly killed, but not too regularly that the (generally quite slow) disk I/O slows down performance too much.
+
+- A reindex, amongst other things, recreates the block index from the block files on disk. The block files themselves don't need to be updated: blocks are already assumed to not be stored sequentially in or across block files anyway.
+
+- This PR skips flushing to disk those blocks that already have a known position in the datastore. Skipping this means that users can write-protect the blk files on disk which may be useful for security or even safely sharing that data between multiple bitcoind instances.
+
+- Some of the functions that are relevant to the reindexing process behave differently than can perhaps be expected from their name:
+  - `SaveBlockToDisk()` does not actually write anything to disk when (as is the case for a reindex) the known block position is passed (as `dbp`), but we still need to call it to update the blockfile info stats
+  - `FindBlockPos()` is mostly used to find the position in a blockfile where a new block is going to be appended. However, it also updates the blockfile info and may also trigger the flushing of block and undo files, which is why we still call it even if we already know the block's position.
+  - this behaviour arguably could arguably be improved upon, as for example done in [this WIP by mzumsande](https://github.com/bitcoin/bitcoin/compare/master...mzumsande:bitcoin:202207_refactor_findblockpos)
 
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+1. After this PR, can the whole `blocks/` directory be made read-only? Why (not)?
+
+1. In [`FindBlockPos()`](https://github.com/bitcoin/bitcoin/blob/19764dc143281376ea08e954018479ed10405b72/src/node/blockstorage.cpp#L611), what is the relevance of `fKnown` to this PR, conceptually? How does the behaviour of `FindBlockPos()` change when `fKnown==true` instead of `false`?
+
+1. What happens if bitcoind is killed after when block(s) have been processed, but before `FlushBlockFile()` is called? Does this PR affect that behaviour?
```

Comment:
This is actually a good question! And actually made writing the test pretty tricky. The reason is because the platform may flush at any time, usually when we `fwrite()`. So it's kinda like `FlushBlockFile()` could be called `EnsureFlush()` like, do it now if it hasn't happened already.